frontend/javascripts/oxalis/model/sagas/annotation_saga.ts

```diff
@@ -159,4 +178,75 @@ export function* watchAnnotationAsync(): Saga<void> {
   );
   yield* takeLatest("EDIT_ANNOTATION_LAYER", pushAnnotationLayerUpdateAsync);
 }
-export default [warnAboutSegmentationZoom, watchAnnotationAsync];
+
+export function* acquireAnnotationMutexMaybe(): Saga<void> {
+  yield* take("WK_READY");
+  const allowUpdate = yield* select((state) => state.tracing.restrictions.allowUpdate);
+  const annotationId = yield* select((storeState) => storeState.tracing.annotationId);
+  if (!allowUpdate) {
+    return;
+  }
+  const othersMayEdit = yield* select((state) => state.tracing.othersMayEdit);
+  const ONE_MINUTE = 1000 * 60;
+  let isInitialRequest = true;
+  let doesHaveMutex = false;
+  let shallTryAcquireMutex = othersMayEdit;
+
+  function onMutexStateChanged(canEdit: boolean, blockedByUser: APIUserCompact | null | undefined) {
+    if (canEdit) {
+      Toast.close("MutexCouldNotBeAcquired");
+      if (!isInitialRequest) {
+        Toast.success(messages["annotation.acquiringMutexSucceeded"]);
+        location.reload();
+      }
+    } else {
+      const message =
+        blockedByUser != null
+          ? messages["annotation.acquiringMutexFailed"]({
+              userName: `${blockedByUser.firstName} ${blockedByUser.lastName}`,
+            })
+          : messages["annotation.acquiringMutexFailed.noUser"];
+      Toast.error(message, { sticky: true, key: "MutexCouldNotBeAcquired" });
+    }
+  }
+
+  function* tryAcquireMutex(): Saga<void> {
+    while (shallTryAcquireMutex) {
+      if (!doesHaveMutex) {
+        yield* put(setAnnotationAllowUpdateAction(false));
+      }
+      try {
+        const { canEdit, blockedByUser } = yield* call(acquireAnnotationMutex, annotationId);
+        yield* put(setAnnotationAllowUpdateAction(canEdit));
+        if (canEdit !== doesHaveMutex || isInitialRequest) {
+          doesHaveMutex = canEdit;
+          onMutexStateChanged(canEdit, blockedByUser);
+        }
+      } catch (error) {
+        console.error("Error while trying to acquire mutex.", error);
+        if (doesHaveMutex === true || isInitialRequest) {
+          onMutexStateChanged(false, null);
+          yield* put(setAnnotationAllowUpdateAction(false));
+        }
+      }
+      isInitialRequest = false;
+      yield* call(delay, ONE_MINUTE);
+    }
+  }
+  if (shallTryAcquireMutex) {
+    yield* fork(tryAcquireMutex);
+  }
+  function* reactToOthersMayEditChanges({
+    othersMayEdit,
+  }: SetOthersMayEditForAnnotationAction): Saga<void> {
+    shallTryAcquireMutex = othersMayEdit;
+    if (shallTryAcquireMutex) {
+      yield* call(tryAcquireMutex);
```

Comment:
I think it can happen that multiple instances of `tryAcquireMutex` are running. Imagine something like this:

- tryAcquireMutex runs because othersMayEdit is true
- the user disables othersMayEdit and re-enables it again quickly (so quickly that the while loop in the saga doesn't terminate; e.g., because it's sleeping in that period)
- now, line 244 will spawn a second instance of tryAcquireMutex

I'd suggest to store a reference of the forked/called saga and then cancel it before starting a new one. See https://redux-saga.js.org/docs/advanced/TaskCancellation for reference.

Reply:
Ah nice, thanks for spotting that and also directly providing a solution 🙏 🚀

Reply:
Do you think I should add another test testing the cancelation mechanism? I am asking because this might take some more time as I already invested quite some time into the tests

Reply:
> Do you think I should add another test testing the cancelation mechanism? I am asking because this might take some more time as I already invested quite some time into the tests

I think, it's not necessary. If each fork/call invocation is accompanied by a cancel call, this seems quite robust to me :)

Reply:
Ok, ss the cancellation is mixed into the saga, the test broke due to this change. Therefore I had to fix the tests and thus also test the cancellation mechanism anyway.
So the canceling is now tested anyway 🤷